### PR TITLE
Documentation: add CoreOS getting-started-guide

### PR DESCRIPTION
**Pull request by ianvernon**
Add a guide with steps to configure Kubernetes and Cilium to get Cilium up-and-running on CoreOS.

Signed-off by: Ian Vernon <ian@covalent.io>

Fixes: #913

### CODE REVIEW

**Comment by ianvernon** [comment]
I tested my guide using a local Vagrant setup and it succeeded, indicating that the guide is accurate.

**Review by aalemayhu — CHANGES_REQUESTED** [review]
Have not tried  following the guide step by step yet. Have to setup
CoreOS with k8s, but will try again later this week.

Some nitpicks from eye balling the commit, PDF and
HTML rendering.

In the commit message
- s/Kubernetesa/Kubernetes/g
- I would also squash the merge commit

Documentation

- Maybe this would look better grouped under 'Getting Started Guide'?
- I would use the `::` frame rendering wherever possible in the guide to
get a consistent look. 
- 'Pre-Setup / Ensure Minimum Requirements Are Met' -> 'Minimum
Requirements'?
- Maybe splitting up 'Step 2: Setup Kubelet' into more pieces like 'Step
2.1 - X', 'Step 2.2 - Y'.  Might read better?

Thanks.

File: Documentation/coreos.rst [hunk 1]
```diff
@@ -0,0 +1,356 @@
+.. _coreos_gs_guide:
+
+Getting Started Guide for CoreOS
+=====================
+
+This document serves as a guide to get Cilium up-and-running on CoreOS.
+
+If you haven't read the :ref:`intro` yet, we encourage you to do that first.
+
+For an introduction about how to setup Kubernetes with CoreOS, we recommend
```
Comment (tgraf) [inline Documentation/coreos.rst:10]: Will be great to cover in a sentence what this guide covers and for what we defer to the CoreOS guide here. The reader should know what to look for in that guide and when to return. Knowing my own reading pattern, I would never come back ;-)

File: Documentation/coreos.rst [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+Slack channel <https://cilium.herokuapp.com>`_. With Cilium contributors
+across the globe, there is almost always someone available to help.
+
+
+Pre-Setup / Ensure Minimum Requirements Are Met
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* Make sure you understand the various `deployment options <https://coreos.com/kubernetes/docs/latest/getting-started.html#deployment-options>`_.
+* *etcd*
+    * Cilium on CoreOS has been tested using etcd as the key-value store for Kubernetes. We recommend using at least version 3.1.0 of etcd. To set up etcd, follow `the CoreOS instructions for setting up an etcd cluster <https://coreos.com/kubernetes/docs/latest/getting-started.html#deploy-etcd-cluster>`_.
```
Comment (tgraf) [inline Documentation/coreos.rst]: We require 3.1.x. We should make this clear.

File: Documentation/coreos.rst [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    * Mount the directory where CNI plugins are installed:
+
+        ::
+
+             --volume cni-bin,kind=host,source=/opt/cni/bin \
+             --mount volume=cni-bin,target=/opt/cni/bin
+
+    * Mount the BPF filesystem so that the information stored there persists across Cilium restarts:
+
+        ``ExecStartPre=/bin/mount bpffs /sys/fs/bpf -t bpf``
```
Comment (tgraf) [inline Documentation/coreos.rst]: Let's refer to here as well for more information:
http://docs.cilium.io/en/latest/admin/#mounting-the-bpf-fs

File: Documentation/coreos.rst [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+
+
+Step 2: Setup Kubelet
+--------------------------------
+
+On CoreOS, Kubelet (the Kubernetes agent that runs on each node, more info `here <https://kubernetes.io/docs/admin/kubelet/>`_) runs as a container. In order to securely communicate with the API server, kubelet uses the TLS assets we generated as part of Step 1.
+
+Master nodes are usually not scheduled to run workloads, so we provide the ``-register-schedulable=false`` in the example YAML excerpt below. This ensures that workloads are only scheduled on worker nodes.
+
+Since we are setting up Kubelet to use Cilium, we want to configure its networking to utilize CNI (Container Networking Interface). This ensures that each pod that is created can communicate with one another within the cluster with Cilium networking configured.
```
Comment (tgraf) [inline Documentation/coreos.rst:43]: Please indicate what is different from the regular CoreOS kubelet configuration and then provide the full example

File: Documentation/coreos.rst [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+   Restart=always
+   RestartSec=10
+
+   [Install]
+   WantedBy=multi-user.target
+
+
+Step 3: Setup kube-apiserver on the master nodes
+--------------------------------
+Follow the `CoreOS instructions for setting up the API server <https://coreos.com/kubernetes/docs/latest/deploy-master.html#set-up-the-kube-apiserver-pod>`_.
```
Comment (tgraf) [inline Documentation/coreos.rst]: "Cilium has no special requirements with regards to the `kube-apiserver` setup. Please follow the `CoreOS instructions...`"

File: Documentation/coreos.rst [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+          path: /etc/kubernetes/ssl
+        name: ssl-certs-kubernetes
+      - hostPath:
+          path: /usr/share/ca-certificates
+        name: ssl-certs-host
+
+
+Step 5: Setup kube-scheduler on the master nodes
+--------------------------------
+Follow the `CoreOS instructions for setting up kube-scheduler <https://coreos.com/kubernetes/docs/latest/deploy-master.html#set-up-the-kube-scheduler-pod>`_.
```
Comment (tgraf) [inline Documentation/coreos.rst]: Same note here as with Step 3.

File: Documentation/coreos.rst [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+.. _cilium-daemonset-deployment:
+
+Step 9: Deploy Cilium DaemonSet
+--------------------------------
+
+If you are using RBAC in your cluster, apply the RBAC ClusterRole, ServiceAccount, and ClusterRoleBinding for Cilium. To do so, use the following command:
+
+``kubectl create -f https://raw.githubusercontent.com/cilium/cilium/master/examples/kubernetes/rbac.yaml``
+
+Download the `Cilium DaemonSet <https://github.com/cilium/cilium/blob/master/examples/kubernetes/cilium-ds.yaml>`_.
```
Comment (tgraf) [inline Documentation/coreos.rst]: We should document how to point it to the etcd of Kubernetes instead.

**Review by tgraf — CHANGES_REQUESTED** [review]
This looks awesome overall. Some comments inline.

File: Documentation/coreos.rst (hunk 4 above)
Comment (ianvernon) [reply Documentation/coreos.rst:43]: I think that the section, `Cilium-specific configuration` covers what differs from the regular CoreOS Kubelet configuration. What do you mean by 'provide the full example'? There is a full example YAML file as part of this step.

File: Documentation/coreos.rst (hunk 1 above)
Comment (ianvernon) [reply Documentation/coreos.rst:10]: Done.

File: Documentation/coreos.rst (hunk 2 above)
Comment (ianvernon) [reply Documentation/coreos.rst]: Done.

File: Documentation/coreos.rst (hunk 3 above)
Comment (ianvernon) [reply Documentation/coreos.rst]: Done.

File: Documentation/coreos.rst (hunk 5 above)
Comment (ianvernon) [reply Documentation/coreos.rst]: Done.

File: Documentation/coreos.rst (hunk 6 above)
Comment (ianvernon) [reply Documentation/coreos.rst]: Done.

File: Documentation/coreos.rst (hunk 7 above)
Comment (ianvernon) [reply Documentation/coreos.rst]: Will address in next commit.

File: Documentation/coreos.rst [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+
+Please make sure that the location provided for ``--k8s-kubeconfig-path`` as an argument to Cilium in the YAML file matches the mounts and that the kubeconfig file exists on each host machine in your cluster at the path specified for mounting (e.g., ``~/.kube/config``).
+
+Setup Worker Nodes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Step 1: Create TLS Assets
+--------------------------------
+iCilium has no special requirements for setting up the TLS assets on worker nodes. Follow the `CoreOS instructions for creating TLS assets on worker nodes<https://coreos.com/kubernetes/docs/latest/deploy-workers.html#tls-assets>`_ for secure communication with the ``kube-apiserver``.
+
```
Comment (aalemayhu) [inline Documentation/coreos.rst]: s/iCilium/Cilium/g, and this link will not render correctly.

Comment (ianvernon) [reply Documentation/coreos.rst]: Good catch.

**Comment by ianvernon** [comment]
@scanf I think splitting up the Kubelet Setup into multiple steps is overkill. I'm open to moving this to be part of the GSG. Once I've addressed all changes I will squash the commits and fix the typo in the commit message.

File: Documentation/coreos.rst (hunk 4 above)
Comment (tgraf) [reply Documentation/coreos.rst:43]: Yes, the full example is great. I think it makes sense to indicate what is different from the stock CoreOS `kubelet.service` to help users understand what they will run differently.

Comment (tgraf) [reply Documentation/coreos.rst:43]: OK, I now understand the structure. I did not understand that `Cilium-specific configuration` serves as a header for the paragraphs following it.

**Review by tgraf — APPROVED** [review]

File: examples/kubernetes/cilium-ds-gsg.yaml [hunk 9]
```diff
@@ -0,0 +1,198 @@
+kind: ClusterRole
```
Comment (ianvernon) [inline examples/kubernetes/cilium-ds-gsg.yaml]: I added this file to be used during the k8s-gsg-test.sh as we cannot use etcd as the key-value store for Cilium in the GSG test because the test is using etcd version 3.0.17, and there currently is no image `>= 3.1.0` for etcd in the container registry used by kubeadm for [etcd](https://console.cloud.google.com/gcr/images/google-containers/GLOBAL/etcd-amd64)

File: Documentation/coreos.rst [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+--------------------------------
+
+On CoreOS, Kubelet (the Kubernetes agent that runs on each node, more info `here <https://kubernetes.io/docs/admin/kubelet/>`_) runs as a container. In order to securely communicate with the API server, kubelet uses the TLS assets we generated as part of Step 1.
+
+Master nodes are usually not scheduled to run workloads, so we provide the ``-register-schedulable=false`` in the example YAML excerpt below. This ensures that workloads are only scheduled on worker nodes.
+
+Since we are setting up Kubelet to use Cilium, we want to configure its networking to utilize CNI (Container Networking Interface). This ensures that each pod that is created can communicate with one another within the cluster with Cilium networking configured.
+
+* Replace ${ADVERTISE_IP} with this node's publicly routable IP.
+* Replace ${DNS_SERVICE_IP}
```
Comment (aanm) [inline Documentation/coreos.rst]: Replace with what? Let's assume this is the first time the user is seeing this so also tell the user that should be an IP from the service-cidr (SERVICE_CLUSTER_IP_RANGE) set bellow

File: Documentation/coreos.rst [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+     --volume var-log,kind=host,source=/var/log \
+     --mount volume=var-log,target=/var/log \
+     --volume dns,kind=host,source=/etc/resolv.conf \
+     --mount volume=dns,target=/etc/resolv.conf \
+     --volume cni-bin,kind=host,source=/opt/cni/bin \
+     --mount volume=cni-bin,target=/opt/cni/bin \
+     --volume etc-cni,kind=host,source=/etc/cni/net.d \
+     --mount volume=etc-cni,target=/etc/cni/net.d"
+   ExecStartPre=/usr/bin/mkdir -p /etc/cni/net.d
+   ExecStartPre=/bin/mount bpffs /sys/fs/bpf -t bpf
```
Comment (aanm) [inline Documentation/coreos.rst]: Can you add a comment why is this here? And / or write a comment before this kubelet.service so users know this is needed for cilium?

Also this is more correct:
```
ExecStartPre=/bin/bash -c ' \\
 if [[ \$(/bin/mount | /bin/grep /sys/fs/bpf -c) -eq 0 ]]; then \\
   /bin/mount bpffs /sys/fs/bpf -t bpf; \\
 fi'
```

otherwise the user might end up with multiple mounts of the bpffs whenever does `sudo service kubelet restart`

File: Documentation/coreos.rst [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+
+    quay.io/coreos/hyperkube:v${KUBE_VERSION}_coreos.0
+
+Step 4: Setup kube-controller-manager on the master nodes
+------------------------------------------------------------
+Per the `CoreOS guide <https://coreos.com/kubernetes/docs/latest/deploy-master.html#set-up-the-kube-controller-manager-pod>`_, "the controller manager is responsible for reconciling any required actions based on changes to `Replication Controllers <https://coreos.com/kubernetes/docs/latest/replication-controller.html>`_. For example, if you increased the replica count, the controller manager would generate a scale up event, which would cause a new Pod to get scheduled in the cluster. The controller manager communicates with the API to submit these events.
+
+Create `/etc/kubernetes/manifests/kube-controller-manager.yaml`. It will use the TLS certificate placed on disk earlier."
+
+* Add ``--allocate-node-cidrs`` to ensure that the kube-controller-manager allocates unique CIDR blocks for each node in the cluster.
```
Comment (aanm) [inline Documentation/coreos.rst]: `allocates unique CIDR` -> `allocates unique pod CIDR`

File: Documentation/coreos.rst [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+Follow `the CoreOS instructions to download kubectl <https://coreos.com/kubernetes/docs/latest/configure-kubectl.html#download-the-kubectl-executable>`_.
+
+* Replace ${MASTER_HOST} with the master node address or name used in previous steps
+* Replace ${CA_CERT} with the absolute path to the ca.pem created in previous steps
+* Replace ${ADMIN_KEY} with the absolute path to the admin-key.pem created in previous steps
+* Replace ${ADMIN_CERT} with the absolute path to the admin.pem created in previous steps
+
+:: 
+
+    kubectl config set-cluster default-cluster --server=https://${MASTER_IP} --certificate-authority=${CA_CERT} --embed-certs=true
```
Comment (aanm) [inline Documentation/coreos.rst:242]: Small nit don't you want to use the `--kubeconfig=cilium.kubeconfig` to so that this one will be specific for cilium?

File: Documentation/coreos.rst [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+        ::
+
+            --volume cni-bin,kind=host,source=/opt/cni/bin \
+            --mount volume=cni-bin,target=/opt/cni/bin 
+
+    * `Mount the BPF filesystem <http://docs.cilium.io/en/latest/admin/#mounting-the-bpf-fs>`_ so that the information stored there persists across Cilium restarts:
+
+        ::  
+
+            ExecStartPre=/bin/mount bpffs /sys/fs/bpf -t bpf
```
Comment (aanm) [inline Documentation/coreos.rst]: Same note as above

File: Documentation/coreos.rst [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+Step 2: Setup Kubelet
+--------------------------------
+
+On CoreOS, Kubelet (the Kubernetes agent that runs on each node, more info `here <https://kubernetes.io/docs/admin/kubelet/>`_) runs as a container. In order to securely communicate with the API server, kubelet uses the TLS assets we generated as part of Step 1.
+
+Since we are setting up Kubelet to use Cilium, we want to configure its networking to utilize CNI (Container Networking Interface). This ensures that each pod that is created can communicate with one another within the cluster with Cilium networking configured.
+
+* Replace ${MASTER_HOST}
+* Replace ${ADVERTISE_IP} with this node's publicly routable IP.
+* Replace ${DNS_SERVICE_IP}
```
Comment (aanm) [inline Documentation/coreos.rst]: Same note as above

File: Documentation/coreos.rst [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+   Environment="RKT_RUN_ARGS=--uuid-file-save=/var/run/kubelet-pod.uuid \
+    --volume var-log,kind=host,source=/var/log \
+    --mount volume=var-log,target=/var/log \
+    --volume dns,kind=host,source=/etc/resolv.conf \
+    --mount volume=dns,target=/etc/resolv.conf \
+    --volume cni-bin,kind=host,source=/opt/cni/bin \
+    --mount volume=cni-bin,target=/opt/cni/bin \
+    --volume etc-cni,kind=host,source=/etc/cni/net.d \
+    --mount volume=etc-cni,target=/etc/cni/net.d"
+   ExecStartPre=/bin/mount bpffs /sys/fs/bpf -t bpf
```
Comment (aanm) [inline Documentation/coreos.rst]: Same note as above

File: Documentation/coreos.rst (hunk 11 above)
Comment (ianvernon) [reply Documentation/coreos.rst]: We do have an explanation above which points to the documentation about mounting the bpffs, so I don't see a need for a comment here.

I'll change the statement to the code you have above to avoid the issue with `kubelet restart` that you mentioned.

File: Documentation/coreos.rst (hunk 13 above)
Comment (ianvernon) [reply Documentation/coreos.rst:242]: What is the need to have this specific to Cilium?

Comment (aanm) [reply Documentation/coreos.rst:242]: So that we have a kubeconfig dedicated for cilium without copying (possible) sensitive data from the default kubeconfig where users store all kubernetes clusters.

**Review by aanm — APPROVED** [review]

**Comment by tgraf** [comment]
Are we waiting on something here?

**Comment by ianvernon** [comment]
@tgraf I am going to merge this then address some of the nits separately which are not related to the content of the guide (e.g., its location in the repository).

**Review by aalemayhu — APPROVED** [review]

**Comment by ianvernon** [comment]
Created https://github.com/cilium/cilium/issues/1119 to track moving to the GSG file.

File: Documentation/coreos.rst (hunk 13 above)
Comment (ianvernon) [reply Documentation/coreos.rst:242]: Will do as part of #1119.
